### ISSUE TITLE
Don't lose commit-related comments on rebase

### ISSUE THREAD

**Issue by GrahamTheCoder** [isaacs/github#443]
Problem: If I rebase my pull request to track the latest master, the hashes change, so any comments on the old commits are no longer visible on the pull request. Sometimes PRs take a long time to merge, and it's very useful to be able to rebase and keep comments.

Implementation: When commenting on a commit currently in a pull request, that comment should be saved permanently to the pull request conversation with a message like "formerly part of this PR"

If it's easier: When that commit is no longer in the branch the PR tracks, show the comment the same way as you show other "outdated diffs".

**Comment by cirosantilli** [comment]
Why not comment on the pull request instead?

**Comment by GrahamTheCoder** [comment]
They aren't my comments, they're other peoples, it'll be hard to get everyone to consistently ignore the comment on commit feature.
Why allow commenting on commits in the first place if the comments will be lost?
Wouldn't it be better to just allow commenting on branches before they're opened as pull requests?

**Comment by MHova** [comment]
My team prefers to comment on the commits themselves during PR reviews because we like to read PRs commit-by-commit so that we can follow along with the author's thought process. We also prefer rebasing to merging and we find it incredibly painful to lose commit comments when rebasing.

Would it be possible to automatically attach commit comments to the PR itself? And if a line in a commit will be changed by a later commit in the same PR, the reviewer can be shown a warning before s/he writes a comment on that line?

**Comment by codemac** [comment]
This is incredibly frustrating. Github really encourages this style of review (many commits in a pr showing progress over time) but many projects dislike having the history of review meddle with their ability to git-bisect.

**Comment by mdentremont** [comment]
+1 (@MHova that describes our situation as well)

**Comment by evanrmurphy** [comment]
This is a real gotcha with rebasing pull requests on GitHub. Would love to see this issue addressed.

**Comment by nvachhar** [comment]
In our team's workflow, we have code review comments addressed by adding new commits to the existing branch (and PR). Reviewing only what's changed since the last time you reviewed a change is useful in our flow. I'd even love being able to comment on a range of commits (i.e., all the commits since the last time I've reviewed), but not losing comments on rebase would be a huge step for us.

**Comment by donquixote** [comment]
Two solutions I can see:
1. Leave the original PR unchanged, close it, open a new PR.
2. Create a tag at the original top of the PR, then rebase. The tag makes sure that comments are not deleted.

I have not tried this second solution with github. But I imagine it should work..

**Comment by babineaum** [comment]
@GrahamTheCoder Hey looks like some of the aspects in your request here were implemented in the new "review comments" feature on github. We noticed now that in-line comments are no longer removed after being addressed with a new commit and are kept in the history on the "Conversation" tab, with the one minor(but glaring) issue that it doesn't mark the old comment in anyway like you suggested with "formerly part of this PR". We are feeling the frustration of the lack of documentation on these new "review comments" changes and how they were implemented. We would like to join the fray and request some feedback be heard from the community. At the conference last week (great time) we heard that they are "listening" and that they are going to do a better job to publish their public roadmap. They are going to be vague but at least get the idea across of what they are working on in the future.

**Comment by GrahamTheCoder** [comment]
Yes I did notice that review comments are now immutable parts of the conversation, whereas line comments are not.

**Comment by babineaum** [comment]
Hey Graham,
I actually think that the line comments are also immutable now too.

**Comment by GrahamTheCoder** [comment]
Ah, I haven't actually tested them in a rebase situation. I just noticed you couldn't edit/delete review comments but could still edit/delete line comments.

**Comment by jamiesalts** [comment]
Comments not going away when a commit changes that line is incredibly frustrating.

**Comment by carlescufi** [comment]
+1 for this one, our [project](https://github.com/zephyrproject-rtos/zephyr) uses a rebase-driven PR approach as well. Supporting `push -f` properly so that rebased PRs display their history would be a major help for us.

**Comment by aspiers** [comment]
IIUC this issue is a symptom of the broader issue that [amending PR branches via force-push causes unreachable commits to be garbage-collected (#997)](https://github.com/isaacs/github/issues/997).  It's not just during rebase, but any force-push.

**Comment by carlescufi** [comment]
@aspiers it is indeed. I just mentioned the rebase-driven approach because typically you either "pile-up" commits in the PR or you rebase (or amend if it's a single commit), and it's the latter that doesn't play nice with GitHub today.
